src/radical/pilot/agent/scheduler/continuous.py

```diff
@@ -175,81 +175,81 @@ def _find_resources(self, node, find_slots, ranks_per_slot, cores_per_slot,
         '''
 
-      # self._log.debug('find on %s: %s * [%s, %s]', node['uid'], )
+        slots = list()
 
-        # check if the node can host the request
-        free_cores = node['cores'].count(rpc.FREE)
-        free_gpus  = node['gpus'].count(rpc.FREE)
-        free_lfs   = node['lfs']
-        free_mem   = node['mem']
+        # find at most `n_slots`
+        while len(slots) < n_slots:
 
-        # check how many slots we can serve, at most
-        alc_slots = 1
-        if cores_per_slot:
-            alc_slots = int(m.floor(free_cores / cores_per_slot))
+            node_idx  = node['index']
+            node_name = node['name']
 
-        if gpus_per_slot:
-            alc_slots = min(alc_slots, int(m.floor(free_gpus / gpus_per_slot)))
+            self._log.debug_9('find resources on %s:%d', node_name, node_idx)
+            self._log.debug_9('node: %s', pprint.pformat(node))
+            self._log.debug_9('cps : %s', cores_per_slot)
 
-        if lfs_per_slot:
-            alc_slots = min(alc_slots, int(m.floor(free_lfs / lfs_per_slot)))
+            slot  = {'node_name' : node_name,
+                     'node_index': node_idx,
+                     'cores'     : list(),
+                     'gpus'      : list(),
+                     'lfs'       : lfs_per_slot,
+                     'mem'       : mem_per_slot}
 
-        if mem_per_slot:
-            alc_slots = min(alc_slots, int(m.floor(free_mem / mem_per_slot)))
+            for core_idx,core in enumerate(node['cores']):
 
-        # is this enough?
-        if not alc_slots:
-            return None
+                if core == rpc.FREE:
+                    slot['cores'].append(core_idx)
 
-        if not partial:
-            if alc_slots < find_slots:
-                return None
+                if len(slot['cores']) == cores_per_slot:
+                    break
 
-        # find at most `find_slots`
-        alc_slots = min(alc_slots, find_slots)
+            if len(slot['cores']) < cores_per_slot:
+                self._log.debug_9('not enough cores on %s', node_name)
+                break
 
-        # we should be able to host the slots - dig out the precise resources
-        slots     = list()
-        node_id   = node['node_id']
-        node_name = node['node_name']
+            # gpus can be shared, so we need proper resource tracking.  If
+            # a slot requires one or more GPUs, GPU sharing is disabled.
+            if gpus_per_slot >= 1.0:
 
-        core_idx  = 0
-        gpu_idx   = 0
+                tmp = int(gpus_per_slot)
+                if tmp != gpus_per_slot:
+                    raise ValueError('cannot share GPUs>1')
+                gpus_per_slot = tmp
 
-        for _ in range(alc_slots):
+                for gpu_idx,gpu in enumerate(node['gpus']):
 
-            cores = list()
-            gpus  = list()
+                    if gpu == rpc.FREE:
+                        slot['gpus'].append([gpu_idx, 1.0])
```

Comment:
do we append `_RO`? (kinda what expected)

Reply:
Yeah, we should.  Thanks.